chore: Include Features, Segments in evaluation context schema

khvn26 opened the pull request:
Thanks for submitting a PR! Please check the boxes below:

- [ ] I have added information to `docs/` if required so people know about the feature!
- [ ] I have filled in the "Changes" section below?
- [ ] I have filled in the "How did you test this code" section below?
- [ ] I have used a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) title for this Pull Request

## Changes

Closes https://github.com/Flagsmith/flagsmith/issues/5808.

This PR adds:
- An `EvaluationResult` schema that includes flag and used segment information.
- `segments` and `features` keys to the `EvaluationContext` schema to have all data required for evaluation in a single DTO.

The design goal was to have a structure that is extensible, simple enough to parse and read through, and that provides all required engine data, excluding any data unrelated to engine.

Some new vocabulary is added here that, IMO, make sense for evaluation:
- String `"key"` fields are preferred to numeric `id` fields to more clearly state their purpose.
- A `"variants"` key is used to represent multivariate feature states, where each variant includes a `"weight"` indicating its distribution percentage. I considered using a unified `"values"` array for all feature states — including non-multivariate ones (with a single `{ "weight": 100 }` entry) — but opted against it to avoid ambiguous semantics around control values and the need to validate against empty arrays.

Both objects will be used by the new engine interface to be introduced in https://github.com/Flagsmith/flagsmith-engine/issues/234.

An example `EvaluationContext` object:

```json
{
  "environment": {
    "key": "abcde",
    "name": "Test Environment"
  },
  "features": {
    "feature_1": {
      "key": "1",
      "name": "feature_1",
      "enabled": true,
      "value": "foobar"
    },
    "feature_2": {
      "key": "2",
      "name": "feature_2",
      "enabled": false,
      "value": "control",
      "variants": [
        {
          "value": "a",
          "weight": 51.0
        },
        {
          "value": "b",
          "weight": 49.0
        }
      ]
    }
  },
  "segments": {
    "test_segment": {
      "key": "1",
      "name": "test_segment",
      "rules": [
        {
          "type": "ALL",
          "sub_rules": [
           {
          "type": "ANY",
          "conditions": [
            {
              "path": "$.identity.identifier",
              "operator": "IN",
              "value": "id1,id2"
            },
            {
              "path": "some-trait",
              "operator": "EQUALS",
              "value": "foobar"
            }
          ]
        }
          ]
        }
      ]
    }
  }
}
```

An example `EvaluationResult` object:

```json
{
  "context": {...},
  "flags": [
    {
      "enabled": false,
      "value": "foobar",
      "feature": {
        "key": "1",
        "name": "feature_1"
      }
    },
    {
      "enabled": true,
      "value": null,
      "feature": {
        "key": "2",
        "name": "feature_2"
      }
    }
  ],
  "segments": [
    {
      "key": "1",
      "name": "test_segment"
    }
  ]
}
```


## How did you test this code?

N/A

Zaimwa9 reviewed (commented):
Looks solid overall. One piece feels missing to make the evaluation fully representative imo - but let me know if it's really useful. 

Right now we have:
- the flag being evaluated
- the identity under evaluation
- the segments that can override
- the final evaluation result

What could be missing in the result is the why do I have those values. We know which segment overid but we don't have necessarily its values (especially thinking with multi-variate where the possibilities are wider).
Do you think it would make sense to surface the override override configuration (blue 40%/red60% instead of blue80%/Green20%) ?

gagantrivedi commented on sdk/evaluation-context.json:
```diff
@@ -23,6 +23,68 @@
+                    "title": "Key",
+                    "type": "string"
+                },
+                "name": {
```
Are the feature name and environment used in the evaluation, or are they just there to make things more readable?

emyller commented on sdk/evaluation-context.json:
```diff
@@ -23,6 +23,68 @@
+                }
+            },
+            "required": [
+                "key",
```
What if not a multivariate feature?

khvn26 replied:
We could probably change this definition to an `anyOf` and come up with two schemas with different sets of required properties, but I think it'll unnecessarily complicate code generation.

emyller commented on sdk/evaluation-context.json:
```diff
@@ -53,12 +123,121 @@
+                }
+            },
+            "required": [
+                "key",
```
Again, it looks odd that `key` is required given not every segment uses `%` rules.

emyller commented on sdk/evaluation-context.json:
```diff
@@ -53,12 +123,121 @@
+                    "type": "array"
+                },
+                "sub_rules": {
+                    "items": {
+                        "$ref": "#/$defs/SegmentRule"
+                    },
```
Do we have to, or can we, ensure there's a depth limit to recursion of this field?

emyller replied:
This was also recursive in the SegmentSerializer until recently. In order to prevent it, here's the [strategy I took](https://github.com/Flagsmith/flagsmith/blob/4018f07d423fffde133fd6512e06c52f7ade2e37/api/segments/serializers.py#L47-L88).

I agree there is little gain here as the serializer now refuses more than one level of rule nesting, so the improvement is optional.

emyller commented on sdk/evaluation-context.json:
```diff
@@ -53,12 +123,121 @@
+        "SegmentCondition": {
+            "description": "Represents a condition within a segment rule for feature flag evaluation.",
+            "properties": {
+                "path": {
```
Maybe ["property"](https://github.com/Flagsmith/flagsmith/blob/635dd77bce207dc6a8486f776f5327d732ccba91/api/segments/models.py#L313)?

emyller replied:
Can you give an example of `x-flagsmith-engine-path` that can be referenced here?

khvn26 replied:
See one for the Environment name, for example: https://github.com/Flagsmith/flagsmith/blob/b3e6a55903269a0d675ebb9862629944fb30bd3d/sdk/evaluation-context.json#L16-L17

emyller replied:
Sorry, I did not ask the question correctly — I might lack the full context. I'll reach out internally as the discussion will likely be irrelevant to the PR, and resolve or add to this thread later. Thanks though for the answers!

emyller commented on sdk/evaluation-context.json:
```diff
@@ -39,12 +101,20 @@
-                    "additionalProperties": {
-                        "type": "string"
-                    },
+                    "anyOf": [
+                        {
+                            "additionalProperties": {
+                                "type": "string"
+                            },
+                            "type": "object"
+                        },
+                        {
+                            "type": "null"
+                        }
+                    ],
```
Thanks for the explanation!

> 2. (...) Hopefully, we're not going to behave differently for null/missing fields, so, perhaps, simply not including a field in required is sufficient to denote that the property is optional.

Agreed, unless `"traits": null` is also valid input.

> 3. The change from an array to a mapping is required (...).

It makes total sense! The type seems to be a mapping already prior to this PR though — I assume it changed recently along with the context values project?

> 4. The string trait value type is not entirely honest here. (...)

Agreed.

In my understanding, here's what we should have, in addition to deleting `"default": null,` below.

```suggestion
                    "additionalProperties": {
                        "type": ["number", "string", "boolean"]
                    },
```

emyller commented on sdk/evaluation-context.json:
```diff
@@ -23,6 +23,67 @@
+                    "type": "array"
+                },
+                "priority": {
+                    "description": "Indicates the priority of the feature context. Higher values indicate higher priority in case when multiple feature contexts for the same feature are applicable",
```
```suggestion
                    "description": "Indicates the priority of the feature context. Higher values indicate higher priority in case when multiple feature contexts for the same feature are applicable.",
```

emyller reviewed (dismissed):
One optional improvement. Looks great regardless!

Zaimwa9 reviewed (dismissed):
looks neat. Let's see it in action!

khvn26 commented:
I had to add some stuff while finalising https://github.com/Flagsmith/flagsmith-engine/pull/239/files:

- I brought back `FeatureContext.name` in c67d7b892521602b7ce8303cbb87308b652c1df4 to simplify serialisation of `FlagResult`s.
- Added `FeatureContext.feature_key` in 04103f3e13cfa9ccba4af0f2142faea8821b4ff9. Feature (not feature state) ids are used to correctly apply overrides, and are surfaced in `FlagResult` objects to maintain backwards compatibility. Additionally, I removed nesting from `FlagResult`.
- Fixed incorrect `priority` field documentation in f50fb30205bb60d5d8760e53276fc46b60c9ee58.
